plan/logical_plan_test.go

```diff
@@ -1632,3 +1632,106 @@ func checkVisitInfo(c *C, v1, v2 []visitInfo, comment CommentInterface) {
 		c.Assert(v1[i], Equals, v2[i], comment)
 	}
 }
+
+func (s *testPlanSuite) TestTopNPushDown(c *C) {
+	UseDAGPlanBuilder = true
+	defer func() {
+		testleak.AfterTest(c)()
+		UseDAGPlanBuilder = false
+	}()
+	cases := []struct {
+		sql  string
+		best string
+	}{
+		// Test Limit + Agg + Proj .
+		{
+			sql:  "select a, count(b) from t group by b limit 5",
+			best: "DataScan(t)->Aggr(count(test.t.b),firstrow(test.t.a))->Limit->Projection",
+		},
+		// Test TopN + Agg + Proj .
+		{
+			sql:  "select a, count(b) from t group by b order by c limit 5",
+			best: "DataScan(t)->Aggr(count(test.t.b),firstrow(test.t.a),firstrow(test.t.c))->Sort + Limit(5) + Offset(0)->Projection->Projection",
+		},
+		// Test TopN + Join + Proj.
+		{
+			sql:  "select * from t, t s order by t.a limit 5",
+			best: "Join{DataScan(t)->DataScan(s)}->Sort + Limit(5) + Offset(0)->Projection",
+		},
+		// Test Limit + Join + Proj.
+		{
+			sql:  "select * from t, t s limit 5",
+			best: "Join{DataScan(t)->DataScan(s)}->Limit->Projection",
+		},
+		// Test TopN + Left Join + Proj.
+		{
+			sql:  "select * from t left outer join t s on t.a = s.a order by t.a limit 5",
+			best: "Join{DataScan(t)->Sort + Limit(5) + Offset(0)->DataScan(s)}(test.t.a,s.a)->Sort + Limit(5) + Offset(0)->Projection",
```

Comment:
The later `Sort + Limit(5) + Offset(0)` can be changed to `Limit(5) + Offset(0)`?